lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,413 @@
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+/// Regex for matching valid field names. Must contain only word chars, periods and dashes.
+static VALID_FIELD_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[\w\.\-]*$").unwrap());
+
+static MISSING_FIELD_STR: &str = "LogEvent does not contain required field.";
+
+static INVALID_TYPE_STR: &str = "LogEvent contains a value with an invalid type.";
+
+static INVALID_FIELD_NAME_STR: &str = "LogEvent contains a value with an invalid type.";
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer {
+    conformed_log: Option<LogEvent>,
+}
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer {
+            conformed_log: None,
+        }
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            Event::Metric(_) | Event::Trace(_) => {
+                panic!("GELF Serializer does not support Metric or Trace events.")
+            }
+        }
+    }
+
+    /// Validates that the GELF required fields exist in the event.
+    fn validate_required_fields(&mut self, log: &LogEvent) -> vector_core::Result<()> {
+        // emits the GelfSerializeFailedMissingField internal event and returns Err
+        fn emit_missing_field(field: &str) -> vector_core::Result<()> {
+            vector_core::internal_event::emit(SerializeFailedMissingField {
+                format_type: FORMAT_TYPE_GELF,
+                message: MISSING_FIELD_STR,
+                field,
+            });
+            Err(format!("{}: {}", MISSING_FIELD_STR, field).into())
+        }
+
+        if !log.contains(VERSION) {
+            emit_missing_field(VERSION)?;
```

Comment:
I wonder if version is something that the encoder should set since we are controlling the GELF encoding here?

Reply:
That is a good point, for cases where events received were not already GELF.